src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,330 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Controls;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.OLE.Interop;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Newtonsoft.Json.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineControl.xaml
+    /// </summary>
+    internal partial class DocumentOutlineControl : UserControl, IOleCommandTarget
+    {
+        [MemberNotNullWhen(true, nameof(SymbolsTreeItemsSource))]
+        private bool SymbolTreeInitialized { get; set; }
+        private List<DocumentSymbolViewModel>? SymbolsTreeItemsSource { get; set; }
+
+        private ITextSnapshot? Snapshot { get; set; }
+        private IWpfTextView TextView { get; set; }
+
+        private readonly string? _filePath;
+        private ResettableDelay? _delay;
+        private readonly IAsynchronousOperationListener _asyncListener;
+
+        public DocumentOutlineControl(IWpfTextView textView, ITextBuffer textBuffer, ILanguageServiceBroker2 languageServiceBroker, IThreadingContext threadingContext, IAsynchronousOperationListener asyncListener)
+        {
+            InitializeComponent();
+
+            _asyncListener = asyncListener;
+            TextView = textView;
+            _filePath = GetFilePath(textView);
+            Snapshot = textBuffer.CurrentSnapshot;
+            SymbolTreeInitialized = false;
+
+            threadingContext.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await UpdateAsync().ConfigureAwait(false);
+            }).FileAndForget("Document Outline: Active Document Changed");
```

Comment:
`File and forget`?

Reply:
generally speaking, when we kick off async work, we should let the asynclistener know.  that way any integration tests that want to validate state will wait until this work completes.

Reply:
not really clear why JTF is needed here.  Unless it's the case that the control is not created on the UI thread.